File: backend/app/schemas/notification.py
from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.core.validators import MAX_DESCRIPTION_LENGTH, MAX_NAME_LENGTH, CamelCaseModel, sanitize_string
from app.models.notification import NotificationCategory, UrgencyLevel


class NotificationBase(BaseModel):
    category: NotificationCategory
    urgency: UrgencyLevel = UrgencyLevel.MEDIUM
    title: str = Field(min_length=1, max_length=MAX_NAME_LENGTH)
    message: str = Field(min_length=1, max_length=MAX_DESCRIPTION_LENGTH)
    related_entity_type: Optional[str] = Field(default=None, max_length=100)
    related_entity_id: Optional[UUID] = None

    @field_validator('title', 'message', 'related_entity_type', mode='before')
    @classmethod
    def sanitize_text(cls, v: Optional[str]) -> Optional[str]:
        return sanitize_string(v)


class NotificationCreate(NotificationBase):
    pass


class NotificationResponse(CamelCaseModel):
    id: UUID
    user_id: UUID
    category: str
    urgency: str
    title: str
    message: str
    related_entity_type: Optional[str] = None
    related_entity_id: Optional[UUID] = None
    is_read: bool = False
    created_at: datetime
    updated_at: datetime


class NotificationListResponse(CamelCaseModel):
    items: list[NotificationResponse]
    total: int
    limit: int
    offset: int


class UnreadCountResponse(CamelCaseModel):
    unread_count: int

File: frontend/src/types/notification.ts
export type NotificationCategory = 'approval' | 'inspection' | 'defect' | 'update' | 'general'
export type UrgencyLevel = 'low' | 'medium' | 'high' | 'critical'

export interface Notification {
  id: string
  userId: string
  category: NotificationCategory
  urgency: UrgencyLevel
  title: string
  message: string
  relatedEntityType?: string
  relatedEntityId?: string
  isRead: boolean
  createdAt: string
  updatedAt: string
}

export interface NotificationFilters {
  category?: NotificationCategory
  urgency?: UrgencyLevel
  isRead?: boolean
  limit?: number
  offset?: number
}

export interface NotificationListResponse {
  items: Notification[]
  total: number
  limit: number
  offset: number
}

export interface UnreadCountResponse {
  unreadCount: number
}

export type DigestFrequency = 'immediate' | 'daily' | 'weekly'

export interface NotificationPreference {
  id: string
  userId: string
  category: string
  enabled: boolean
  minUrgencyLevel: UrgencyLevel
  quietHoursStart: string | null
  quietHoursEnd: string | null
  emailEnabled: boolean
  pushEnabled: boolean
  digestFrequency: DigestFrequency
  createdAt: string
  updatedAt: string
}

export interface NotificationPreferenceCreate {
  category: string
  enabled?: boolean
  minUrgencyLevel?: UrgencyLevel
  quietHoursStart?: string | null
  quietHoursEnd?: string | null
  emailEnabled?: boolean
  pushEnabled?: boolean
  digestFrequency?: DigestFrequency
}

export interface NotificationPreferenceUpdate {
  category?: string
  enabled?: boolean
  minUrgencyLevel?: UrgencyLevel
  quietHoursStart?: string | null
  quietHoursEnd?: string | null
  emailEnabled?: boolean
  pushEnabled?: boolean
  digestFrequency?: DigestFrequency
}

export interface NotificationPreferenceListResponse {
  items: NotificationPreference[]
  total: number
  limit: number
  offset: number
}
